### PR TITLE
Doorkeeper refresh_token expiration BONNIE-1571

### PR DESCRIPTION
**Pull request by hossenlopp**
Monkey patches the mongoid model for access_tokens to include the time of the original authorization so we can stop the use (expire) of a refresh_token if it has been more than 5 days. The check for 5 days from original token (authorization) time is monkey patched into the request handling for requests to get an access token using the refresh_token. An additional configuration option, refresh_token_expires_in, is monkey patched into the doorkeeper configuration.

Pull requests into Bonnie require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1571
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced. N/A
- [x] Tests are included and test edge cases **Monkey patched code could not be unit tested.**
- [x]  Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Test fixtures updated and documented as necessary ( see [internal wiki](https://gitlab.mitre.org/bonnie/internal-documentation/wikis/testing#test-fixtures) )
- [x] Code coverage has not gone down and all code touched or added is covered. 
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here: 
                 Monkey patched code cannot be unit tested. Manual testing using bonnie-api-client and setting `expires_in` times to `1.minutes` for access_tokens and `2.minutes` refresh_tokens in `config/initializers/doorkeeper.rb`. Is needed.
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here: 
                A JIRA test in our strict sense would take 5 days to complete and also be overly burdensome to develop at this time. We will do manual testing on staging using the short times before deploying.

Branch | Back End Coverage | Front End Coverage
-- | -- | --
bonnie-v2.3 | 1893 / 2124 LOC (89.12%) covered. | Statements   : 59.44% ( 7280/12247 )<br>Branches     : 43.43% ( 1979/4557 )<br>Functions    : 49.81% ( 1207/2423 )<br>Lines        : 57.31% ( 6536/11404 )
...BONNIE-1571 | 1913 / 2155 LOC (88.77%) covered. | Statements   : 59.44% ( 7280/12247 )<br>Branches     : 43.43% ( 1979/4557 )<br>Functions    : 49.81% ( 1207/2423 )<br>Lines        : 57.31% ( 6536/11404 )

- [x] Automated regression test(s) pass N/A does not touch calculation code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA test links: N/A
- [x] Justification for using JIRA tests: N/A
- [x] JIRA tests have been added to sprint N/A

**Reviewer 1:**

Name: @losborne 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [X] <s>JIRA tests have been run and pass</s>
- [X] <s>You agree with the justification for use of JIRA tests or have provided input on why you disagree</s>

**Reviewer 2:**

Name: @c-monkey
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [X] <s>JIRA tests have been run and pass</s>
- [X] <s>You agree with the justification for use of JIRA tests or have provided input on why you disagree</s>

### CODE REVIEW

**Comment by hossenlopp** [comment]
This can be easily tested locally by changing the expires_in times to `1.minutes` for access_tokens and `2.minutes` refresh_tokens in `config/initializers/doorkeeper.rb`

File: lib/ext/doorkeeper/access_token_request.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      # Replaces code in https://github.com/doorkeeper-gem/doorkeeper/blob/v4.2.6/lib/doorkeeper/oauth/refresh_token_request.rb
+      # to support expiration of refresh tokens.
+      ###
+      def before_successful_response
+        refresh_token.transaction do
+          refresh_token.lock!
+          raise Errors::InvalidTokenReuse if refresh_token.revoked?
+
+          # only check for refresh token expiration if it was defined in the config
+          if Doorkeeper.configuration.refresh_token_expires_in
```
Comment (lukoktonos) [inline lib/ext/doorkeeper/access_token_request.rb]: This block is a duplication of  `app/controllers/doorkeeper_override/token_info_controller.rb`.  Is there a way to pull this out into a separate function to remove the duplication?  

Or perhaps it would be better to add `refresh_token_expiration_time` instead of `refresh_expires_in_seconds` to `token_info`, as it is just re-calculated here?

File: test/functional/api_v1/token_info_controller_test.rb [hunk 2]
```diff
@@ -39,6 +40,55 @@ def accessible?
       assert_equal @token.scopes, body['scopes']
       assert_equal @token.expires_in_seconds, body['expires_in_seconds']
       assert_equal @token.created_at.to_i, body['created_at']
+      assert_equal Doorkeeper.configuration.refresh_token_expires_in.to_i - 21, body['refresh_expires_in_seconds']
```
Comment (lukoktonos) [inline test/functional/api_v1/token_info_controller_test.rb]: Does this assert assume 1 second has passed since line 29?  Is there any case where 0 or >1 seconds would pass, thus making the assert fail?

File: test/functional/api_v1/token_info_controller_test.rb [hunk 3]
```diff
@@ -39,6 +40,55 @@ def accessible?
...
+      assert_equal "bonnie", body['user_last_name']
+      assert_equal @token.scopes, body['scopes']
+      assert_equal @token.expires_in_seconds, body['expires_in_seconds']
+      assert_equal @token.created_at.to_i, body['created_at']
+      assert_equal Doorkeeper.configuration.refresh_token_expires_in.to_i - 1, body['refresh_expires_in_seconds']
+    end
+
+    test "successfully shows token when refresh_token expiration is not enabled" do
+      # stash the expiration setting and then change it to nil
+      original_refresh_token_expires_in = Doorkeeper.configuration.refresh_token_expires_in
```
Comment (lukoktonos) [inline test/functional/api_v1/token_info_controller_test.rb:64]: I believe this and the corresponding line 91 belong in a setup/teardown, because the state will not be reset if the test fails before getting to the last line.

**Review by c-monkey — APPROVED** [review]
Tests pass, timeout code works with `Bonnie Api Client` test app

File: test/functional/api_v1/token_info_controller_test.rb (hunk 3 above)
Comment (hossenlopp) [reply test/functional/api_v1/token_info_controller_test.rb:64]: I don't know how to do this for one specific test.

Comment (lukoktonos) [reply test/functional/api_v1/token_info_controller_test.rb:64]: Leaving the first part as-is but putting everything else in a `begin` and the teardown part in an `ensure` block seems to do the trick.

File: lib/ext/doorkeeper/access_token_request.rb (hunk 1 above)
Comment (hossenlopp) [reply lib/ext/doorkeeper/access_token_request.rb]: Pulling out the duplicated functionality may be possible... I'll give it a try. I could put it in the token model.

**Review by lukoktonos — APPROVED** [review]
Looks good.

Manually tested with bonnie-api-client according to @hossenlopp's instructions.  Everything seems to be working correctly.

Agree with decision to repeat manual test on staging, and that it is not feasible to do so on production (since there is a 5 day timeout).

File: test/functional/api_v1/token_info_controller_test.rb (hunk 2 above)
Comment (hossenlopp) [reply test/functional/api_v1/token_info_controller_test.rb]: The rounding of the math I'm doing on the time appears to be different than the rounding the Doorkeeper is doing. It probably isn't worth our time to fix this rounding change.

**Review by c-monkey — APPROVED** [review]
Good code coverage, tests pass.
